[PRELIMINARY REVIEW] Go programming language for backend development (#25)
folio-org/rfcs: text/0009-golang.md

From a sysops perspective, Go containers require less operational tuning compared to the JVM containers as they do not contain a stand-alone virtual machine that requires parametrization.

The Go language is widely used across the industry and open-source communities, especially in the areas related to web and microservices. Popular examples of large open-source projects created with Go include the [Kubernetes](https://github.com/kubernetes/kubernetes) container orchestration platform and the [Docker](https://github.com/docker) container runtime.

The Go project maintains a [list](https://go.dev/solutions/) of case-studies from commercial companies relying on Go in their products and general use-cases for the language.

According to [GitHub statistics](https://madnight.github.io/githut/#/pull_requests/2023/3), Go is the third fastest growing language (behind Python and Java) as measured by the number of new Pull Requests.

> ## Risks and Drawbacks

Author: Yes, good point. I'll add it.  This library can be kept minimal though, I'd use `folio-vertx-lib` as an example of FOLIO integration library that's very small.

> Introducing Go as one of the supported backend technologies involves the same general concerns as introducing any new technology into FOLIO. The biggest potential risk is the increased development and maintenance complexity as some new FOLIO backend modules will be created with Go which is not immediately well known to existing FOLIO backend developers that have primarily Java background. As explained in the previous section, however, this risk is highly minimized because of the design of Go which has many points of similarity to Java.

Reviewer: > As explained in the previous section, however, this risk is highly minimized because of the design of Go which has many points of similarity to Java.

Personal opinion (from someone who likes Go): "highly minimized" feels like a bit of a stretch. Go is a great language, but there are quite a few common idioms that could make some code tricky to follow for someone new to the language.
Here's a short list of things that stand out as things that are pretty radically different from Java (just from a quick perusal of Effective Go), where a Java dev might need to go track down docs just to understand code they are reading:
- `=` vs `:=`
- error handling and `panic`/`recover`
- every loop is a `for`-loop
- `range`
- no fallthrough in `switch` blocks
- non-C-like pointers
- the variable declaration syntax
- functions with naked return statements even though they return values
- deferred functions
- the syntax around type parameters when declaring stuff like arrays and maps
- having to use `make` to allocate some objects, but not others
- slices
- `_` (oh, how I wish Java had `_`...)
- semantics around capitalization
- Go's enums are kind of weird compared to Java's
- `init()`
- `interface {}`
- the standard lib's interfaces can take some getting used to (e.g., `io.Reader` and `io.Writer`)
- goroutines ❤️ 
- channels ❤️ 

(note: this could also be considered a list of features that make Go pretty awesome 🤷)

Author: Well, this is highly subjective but from my personal experience, I think it mostly holds.  The syntax differences you list are of course there but in most cases it's "cosmetics" and not paradigm shifting changes. I'd say the biggest difference is the concurrency model (goroutines) and more advanced topics like generics. Though concurrency is already sufficiently different between traditional thread-base Spring and async/future-based vert.x to make this area more generally complex. Also, pointers and stack vs heap allocation is something that may be unfamiliar to Java programmers, especially with no background in C/C++.

I'll rephrase this to acknowledge some major difference but without getting too much into details.

Additionally, as history of FOLIO development shows, there exists high level of affinity between specific teams of developers and particular modules and domains. This affinity often results in siloing and can create barriers preventing wider collaboration between various development teams. This siloing exists even within Java-only usage in FOLIO (vert.x vs Spring) because of the high reliance on web programming frameworks. Development in Go has less or no reliance on frameworks -- for instance it's entirely possible to develop FOLIO modules with only Go standard library, Postgres client library and OpenAPI parser library. This more lightweight approach to development can ease some of the cross-module development barriers in FOLIO. 

## Rationale and Alternatives

> As explained in the design section, Go is a programming language dedicated to writing backend web services and as such is a perfect fit for FOLIO's microservices-inspired architecture. It is also a language that is increasingly popular with community-driven open source projects thanks to it simplicity and familiar syntax but without the performance tradeoffs that are typical to dynamic or scripting languages.

Reviewer: ```suggestion
As explained in the design section, Go is a programming language well-suited to writing backend web services and, as such, is a perfect fit for FOLIO's microservices-inspired architecture. It is also an increasingly popular language with community-driven open-source projects thanks to its simplicity and familiar syntax but without the performance tradeoffs typical to dynamic or scripting languages.
```

Go's excellent AOT compilation support puts it ahead of Java for building systems that comprise of many small, containerized workloads.

As such there's no direct alternative to Go available in the FOLIO ecosystem.

## Timing

> We propose that Go is added to the list of supported technologies starting from Quesnelia in the form of a pilot project. The pilot will serve as the technology preview with the aim of validating Go's fit for FOLIO backend development. No existing core FOLIO modules will be re-developed using Go during the pilot phase and any new modules written in Go would be considered proof-of-concept modules. 

Reviewer: > No existing core FOLIO modules will be re-developed using Go

Personally, I'd avoid this statement and maybe even consider rewriting one of the smaller modules in Go as a proof-of-concept in order to provide a direct head-to-head comparison of Java vs Go

Author: I agree that rewriting an existing module is a good idea for the PoC and will serve as a good comparison. This para was meant to convey that initially, we will not create dependency between a "core" FOLIO module and go but I see now that it's not clear..

In the case Go is deemed unsuitable for FOLIO development, any modules developed in Go will be removed from the future FOLIO versions and Go will be dropped as a supported technology.

> In the case Go is deemed suitable for FOLIO backend development, it will be accepted to the official technology list without any restrictions starting from Ramsons.

Reviewer: Since this discusses a go/no-go decision for officially supporting the language (😄), it may be worthwhile to propose some timing for the decision
